lib/posix-futex: Clear thread's objects upon thread exit

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

If a thread creates a futex that is not cleared upon application exit that futex will remain in futex_list after the unmap of the thread's stack causing an abort on the next iteration of the list.

Clear thread's objects upon thread exit.

michpappas committed:
lib/posix-futex: Clear thread's objects upon thread exit

If a thread creates a futex that is not cleared upon application exit
that futex will remain in futex_list after the unmap of the thread's
stack causing an abort on the next iteration of the list.

Clear thread's objects upon thread exit.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

mariapana reviewed (approved):
Looks good.
Reviewed-by: Maria Pana <maria.pana4@gmail.com>

mariasfiraiala commented on lib/posix-futex/futex.c:
```diff
@@ -382,14 +382,25 @@ UK_LLSYSCALL_R_DEFINE(pid_t, set_tid_address, pid_t *, tid_ref)
 
-/* Thread exit handler that clears child TID at the stored reference */
-static void pfutex_child_cleartid_term(struct uk_thread *child __unused)
+static void thread_exit_handler(struct uk_thread *child __unused)
```
Is the `__unused` attribute still needed now? As far as I can see `child` is used at line 399.

michpappas replied:
Good catch, thanks! 👍🏼

mariasfiraiala reviewed (commented):
All good on my side, besides a small comment.

flpostolache reviewed (commented):
I think a note to mention that this function is not SMP safe should be added so that when the full SMP support is added, this function would be also changed.
From my understanding, the `list_node` will break if two threads will try to delete two adjacent futexes at the same time on two different cores.

michpappas commented:
> I think a note to mention that this function is not SMP safe should be added so that when the full SMP support is added, > this function would be also changed.
> From my understanding, the list_node will break if two threads will try to delete two adjacent futexes at the same time on two different cores.

@flpostolache thanks for pointing this out. I added a calls to acquire and release `futex_llist_lock` to protect the list. 

I also added a `break` statement once the thread's entry is found, as pointed out during a private discussion with @mogasergiu, and removed `__unused` as pointed out by @mariasfiraiala.

flpostolache reviewed (approved):
Thanks for the changes! Looks good now.

Reviewed-by: Florin Postolache <florin.postolache.of@gmail.com>

mariasfiraiala reviewed (approved):
All good on my side now too, thanks!

Reviewed-by: Maria Sfiraiala <maria.sfiraiala@gmail.com>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
